usr.bin/uucp/uucp/uucp.1
.\" Copyright (c) 1988 The Regents of the University of California.
.\" All rights reserved.
.\"
.\" This module is believed to contain source code proprietary to AT&T.
.\" Use and redistribution is subject to the Berkeley Software License
.\" Agreement and your Software Agreement with AT&T (Western Electric).
.\"
.\"	@(#)uucp.1	6.6 (Berkeley) 4/24/91
.\"
.TH UUCP 1 "April 24, 1991"
.AT 3
.SH NAME
uucp \- unix to unix copy
.SH SYNOPSIS
.B uucp
[
.B \-acCdfmr
] [
.BI \-n user
] [
.BI \-g grade
] [
.BI \-s spool
] [
.BI \-x debug
] source-file....  destination-file
.SH DESCRIPTION
.I Uucp
copies files named by the source-file arguments
to the destination-file argument.
A file name may be a pathname on your machine, or may 
have the form
.IP
system-name!pathname
.LP
where `system-name' is taken from a list of system names
that
.I uucp
knows about.
Shell metacharacters ?*[] appearing in the pathname part
will be expanded on the appropriate system.
.PP
Pathnames may be one of:
.IP (1)
a full pathname;
.IP (2)
a pathname preceded by
.BI ~ user;
where
.I user
is a userid on the specified system
and is replaced by that user's login directory;
.IP (3)
a pathname prefixed by
.BR ~ ,
where
.B ~
is expanded into the system's public directory (usually /var/spool/uucppublic);
.IP (4)
a partial pathname, which is prefixed by the current directory.
.PP
If the result is an erroneous pathname for the remote system,
the copy will fail.
If the destination-file is a directory, the last part of the
source-file name is used.
.ig
If a simple
.I ~user
destination is inaccessible to 
.IR uucp ,
data is copied to a spool directory and the user
is notified by
.IR mail (1).
..
.PP
.I Uucp
preserves execute permissions across the transmission
and gives 0666 read and write permissions (see
.IR chmod (2)).
.PP
The following options are interpreted by
.IR uucp .
.TP
.B \-a
Avoid doing a
.I getwd
to find the current directory.
(This is sometimes used for efficiency.)
.TP 
.B \-c
Use the source file when copying out rather than
copying the file to the spool directory.
(This is the default.)
.TP
.B \-C
Copy the source file to the spool directory and transmit
the copy.
.TP 
.B \-d
Make all necessary directories for the file copy.
(This is the default.)
.TP
.B \-f
Do not make intermediate directories for the file copy.
.TP
.BI \-g grade
.I Grade
is a single letter/number; lower ASCII sequence characters
will cause a job to be transmitted earlier during a particular conversation.
Default is `n'. By way of comparison,
.IR uux (1)
defaults to `A'; mail is usually sent at `C'.
.TP 
.B \-m
Send mail to the requester when the copy is complete.
.TP
.BI \-n user
Notify
.I user
on remote system (i.e., send
.I user
mail) that a file was sent.
.TP
.B \-r
Do not start the transfer, just queue the job.
.TP
.BI \-s spool
Use
.I spool
as the spool directory instead of the default.
.TP
.BI \-x debug
Turn on the debugging at level
.I debug.
.SH FILES
/var/spool/uucp - spool directory
.br
/usr/lib/uucp/* - other data and program files
.SH SEE ALSO
uux(1), mail(1)
.PP
D. A. Nowitz and M. E. Lesk,
.IR "A Dial-Up Network of UNIX Systems" .
.PP
D. A. Nowitz,
.IR "Uucp Implementation Description" .
.SH WARNING
The domain of remotely accessible files can
(and for obvious security reasons, usually should)
be severely restricted.
You will very likely not be able to fetch files
by pathname;
ask a responsible person on the remote system to
send them to you.
For the same reasons you will probably not be able
to send files to arbitrary pathnames.
.SH BUGS
.br
All files received by
.I uucp
will be owned by the uucp administrator (usually UID 5).
.PP
The 
.B \-m
option will only work sending files or receiving
a single file.
(Receiving multiple files specified by special shell
characters ?*[] will not activate
the
.B \-m
option.)
.PP
At present
.I uucp
cannot copy to a system several "hops" away, that is, a command of
the form
.PP
.nf
	uucp myfile system1!system2!system3!yourfile
.fi
.PP
is not permitted. Use
.IR uusend (1)
instead.
.PP
When invoking
.I uucp
from
.IR csh (1),
the `!' character must be prefixed by the `\e' escape to inhibit
.IR csh 's
history mechanism. (Quotes are not sufficient.)
.PP
.I Uucp
refuses to copy a file that does not give read access to ``other'';
that is, the file must have at least 0444 modes.


usr.bin/uucp/uucp/uucp.c
/*-
 * Copyright (c) 1988 The Regents of the University of California.
 * All rights reserved.
 *
 * This module is believed to contain source code proprietary to AT&T.
 * Use and redistribution is subject to the Berkeley Software License
 * Agreement and your Software Agreement with AT&T (Western Electric).
 */

#ifndef lint
char copyright[] =
"@(#) Copyright (c) 1988 The Regents of the University of California.\n\
 All rights reserved.\n";
#endif /* not lint */

#ifndef lint
static char sccsid[] = "@(#)uucp.c	5.10 (Berkeley) 4/24/91";
#endif /* not lint */

#include "uucp.h"
#include <sys/stat.h>
#include "uust.h"

/*
 *	uucp command
 */

int Uid;
char *Ropt = " ";
char Path[100], Optns[10], Ename[MAXBASENAME+1];
char Grade = 'n';
#ifdef DONTCOPY
int Copy = 0;
#else !DONTCOPY
int Copy = 1;
#endif !DONTCOPY
char Nuser[32];
struct timeb Now;

/* variables used to check if talking to more than one system. */
int	xsflag = -1;
char	xsys[MAXBASENAME+1];

long Nbytes = 0;
#define MAXBYTES 50000	/* maximun number of bytes of data per C. file */
#define MAXCOUNT 15	/* maximun number of files per C. file */

main(argc, argv)
int argc;
char **argv;
{
	char *sysfile1, *sysfl2;
	register char *cp;
	char file1[MAXFULLNAME], file2[MAXFULLNAME];
	int avoidgwd = 0, c;
	extern char *optarg;
	extern int optind;

	strcpy(Progname, "uucp");
	uucpname(Myname);
	umask(WFMASK);
	Optns[0] = '-';
	Optns[1] = 'd';
#ifdef DONTCOPY
	Optns[2] = 'c';
#else !DONTCOPY
	Optns[2] = 'C';
#endif !DONTCOPY
	Ename[0] = Nuser[0] = Optns[3] = '\0';
	while((c = getopt(argc, argv, "aCcdfe:g:mn:rs:x:")) != EOF)
		switch(c) {
		case 'a':
			/* efficiency hack; avoid gwd call */
			avoidgwd = 1;
			break;
		case 'C':
			Copy = 1;
			Optns[2] = 'C';
			break;
		case 'c':
			Copy = 0;
			Optns[2] = 'c';
			break;
		case 'd':
			break;
		case 'f':
			Optns[1] = 'f';
			break;
		case 'e':
			strncpy(Ename, optarg, MAXBASENAME);
			break;
		case 'g':
			Grade = *optarg;
			break;
		case 'm':
			strcat(Optns, "m");
			break;
		case 'n':
			sprintf(Nuser, "%.31s", optarg);
			break;
		case 'r':
			Ropt = argv[optind-1];
			break;
		case 's':
			Spool = optarg;
			break;
		case 'x':
			chkdebug();
			Debug = atoi(optarg);
			if (Debug <= 0)
				Debug = 1;
			fprintf(stderr, "DEBUG %d\n", Debug);
			break;
		case '?':
		default:
			fprintf(stderr, "unknown flag %s\n", argv[optind-1]);
			break;
		}

	DEBUG(4, "\n\n** %s **\n", "START");
	if (!avoidgwd) {
		cp = getwd(Wrkdir);
		if (cp == NULL) {
			syslog(LOG_WARNING, "getwd failed");
			cleanup(1);
		}
	}
	if (subchdir(Spool) < 0) {
		syslog(LOG_WARNING, "chdir(%s) failed: %m", Spool);
		cleanup(1);
	}

	Uid = getuid();
	if (guinfo(Uid, User, Path) != SUCCESS) {
		syslog(LOG_WARNING, "Can't find username for uid %d", Uid);
		DEBUG(1, "Using username", "uucp");
		strcpy(User, "uucp");
	}
	DEBUG(4, "UID %d, ", Uid);
	DEBUG(4, "User %s,", User);
	DEBUG(4, "Ename (%s) ", Ename);
	DEBUG(4, "PATH %s\n", Path);
	if (optind > (argc-2)) {
		fprintf(stderr, "usage: uucp [flags] from... to\n");
		cleanup(1);
	}


	/*  set up "to" system and file names  */
	if ((cp = index(argv[--argc], '!')) != NULL) {
		sysfl2 = argv[argc];
		*cp = '\0';
		if (*sysfl2 == '\0')
			sysfl2 = Myname;
		else
			strncpy(Rmtname, sysfl2, MAXBASENAME);
		if (versys(&sysfl2) != 0) {
			fprintf(stderr, "bad system name: %s\n", sysfl2);
			cleanup(1);
		}
		if (Rmtname[0] != '\0')
			strncpy(Rmtname, sysfl2, MAXBASENAME);
		/* block multi-hop requests immediately */
		if (index(cp+1, '!') != NULL) {
			fprintf(stderr, "uucp handles only adjacent sites.\n");
			fprintf(stderr, "Try uusend for multi-hop delivery.\n");
			cleanup(1);
		}
		strcpy(file2, cp + 1);
	} else {
		sysfl2 = Myname;
		strcpy(file2, argv[argc]);
	}
	if (strlen(sysfl2) > MAXBASENAME)
		sysfl2[MAXBASENAME] = '\0';


	/*  do each from argument  */
	while (optind < argc) {
		if ((cp = index(argv[optind], '!')) != NULL) {
			sysfile1 = argv[optind];
			*cp = '\0';
			if (strlen(sysfile1) > MAXBASENAME)
				sysfile1[MAXBASENAME] = '\0';
			if (*sysfile1 == '\0')
				sysfile1 = Myname;
			else
				strncpy(Rmtname, sysfile1, MAXBASENAME);
			if (versys(&sysfile1) != 0) {
				fprintf(stderr, "bad system name: %s\n", sysfile1);
				cleanup(1);
			}
			if (Rmtname[0] != '\0')
				strncpy(Rmtname, sysfl2, MAXBASENAME);
			strcpy(file1, cp + 1);
		} else {
			sysfile1 = Myname;
			strcpy(file1, argv[optind]);
		}
		DEBUG(4, "file1 - %s\n", file1);
		copy(sysfile1, file1, sysfl2, file2);
		optind++;
	}

	clscfile();
	if (*Ropt != '-' && xsflag >= 0)
		xuucico(xsys);
	cleanup(0);
}

cleanup(code)
int code;
{
	logcls();
	rmlock(CNULL);
	if (code)
		fprintf(stderr, "uucp failed. code %d\n", code);
	exit(code);
}


/*
 *	generate copy files
 *
 *	return codes 0  |  FAIL
 */

copy(s1, f1, s2, f2)
register char *s1, *f1, *s2, *f2;
{
	int type, statret;
	struct stat stbuf, stbuf1;
	char dfile[NAMESIZE];
	char file1[MAXFULLNAME], file2[MAXFULLNAME];
	FILE *cfp, *gtcfile();
	char opts[100];

	type = 0;
	opts[0] = '\0';
	strcpy(file1, f1);
	strcpy(file2, f2);
	if (strcmp(s1, Myname) != SAME)
		type = 1;
	if (strcmp(s2, Myname) != SAME)
		type += 2;
	if (type & 01)
		if ((index(f1, '*') != NULL
		  || index(f1, '?') != NULL
		  || index(f1, '[') != NULL))
			type = 4;

	switch (type) {
	case 0:
		/* all work here */
		DEBUG(4, "all work here %d\n", type);
		if (ckexpf(file1))
			 return FAIL;
		if (ckexpf(file2))
			 return FAIL;
		if (stat(subfile(file1), &stbuf) != 0) {
			fprintf(stderr, "can't get file status %s \n copy failed\n",
			  file1);
			return SUCCESS;
		}
		statret = stat(subfile(file2), &stbuf1);
		if (statret == 0
		  && stbuf.st_ino == stbuf1.st_ino
		  && stbuf.st_dev == stbuf1.st_dev) {
			fprintf(stderr, "%s %s - same file; can't copy\n", file1, file2);
			return SUCCESS;
		}
		if (chkpth(User, "", file1) != 0
		  || chkperm(file2, index(Optns, 'd'))
		  || chkpth(User, "", file2) != 0) {
			fprintf(stderr, "permission denied\n");
			cleanup(1);
		}
		if ((stbuf.st_mode & ANYREAD) == 0) {
			fprintf(stderr, "can't read file (%s) mode (%o)\n",
			  file1, (int)stbuf.st_mode);
			return FAIL;
		}
		if (statret == 0 && (stbuf1.st_mode & ANYWRITE) == 0) {
			fprintf(stderr, "can't write file (%s) mode (%o)\n",
			  file2, (int)stbuf.st_mode);
			return FAIL;
		}
		xcp(file1, file2);
		/* With odd umask() might not be able to read it himself */
		(void) chmod(file2, 0666);
		logent("WORK HERE", "DONE");
		return SUCCESS;
	case 1:
		/* receive file */
		DEBUG(4, "receive file - %d\n", type);
		chsys(s1);
		if (file1[0] != '~')
			if (ckexpf(file1))
				 return FAIL;
		if (ckexpf(file2))
			 return FAIL;
		if (chkpth(User, "", file2) != 0) {
			fprintf(stderr, "permission denied\n");
			return FAIL;
		}
		if (Ename[0] != '\0') {
			/* execute uux - remote uucp */
			xuux(Ename, s1, file1, s2, file2, opts);
			return SUCCESS;
		}

		cfp = gtcfile(s1);
		fprintf(cfp, "R %s %s %s %s\n", file1, file2, User, Optns);
		break;
	case 2:
		/* send file */
		if (ckexpf(file1))
			 return FAIL;
		if (file2[0] != '~')
			if (ckexpf(file2))
				 return FAIL;
		DEBUG(4, "send file - %d\n", type);
		chsys(s2);

		if (chkpth(User, "", file1) != 0) {
			fprintf(stderr, "permission denied %s\n", file1);
			return FAIL;
		}
		if (stat(subfile(file1), &stbuf) != 0) {
			fprintf(stderr, "can't get status for file %s\n", file1);
			return FAIL;
		}
		if ((stbuf.st_mode & S_IFMT) == S_IFDIR) {
			fprintf(stderr, "directory name illegal - %s\n",
			  file1);
			return FAIL;
		}
		if ((stbuf.st_mode & ANYREAD) == 0) {
			fprintf(stderr, "can't read file (%s) mode (%o)\n",
			  file1, (int)stbuf.st_mode);
			return FAIL;
		}
		if ((Nuser[0] != '\0') && (index(Optns, 'n') == NULL))
			strcat(Optns, "n");
		if (Ename[0] != '\0') {
			/* execute uux - remote uucp */
			if (Nuser[0] != '\0')
				sprintf(opts, "-n%s", Nuser);
			xuux(Ename, s1, file1, s2, file2, opts);
			return SUCCESS;
		}
		Nbytes +=  stbuf.st_size;
		if (Copy) {
			gename(DATAPRE, Myname, Grade, dfile);
			if (xcp(file1, dfile) != 0) {
				fprintf(stderr, "can't copy %s\n", file1);
				return FAIL;
			}
		}
		else {
			/* make a dummy D. name */
			/* cntrl.c knows names < 6 chars are dummy D. files */
			strcpy(dfile, "D.0");
		}
		cfp = gtcfile(s2);
		fprintf(cfp, "S %s %s %s %s %s %o %s\n", file1, file2,
			User, Optns, dfile, (int)stbuf.st_mode & 0777, Nuser);
		break;
	case 3:
	case 4:
		/*  send uucp command for execution on s1  */
		DEBUG(4, "send uucp command - %d\n", type);
		chsys(s1);
		if (strcmp(s2,  Myname) == SAME) {
			if (ckexpf(file2))
				 return FAIL;
			if (chkpth(User, "", file2) != 0) {
				fprintf(stderr, "permission denied\n");
				return FAIL;
			}
		}
		if (Ename[0] != '\0') {
			/* execute uux - remote uucp */
			xuux(Ename, s1, file1, s2, file2, opts);
			return SUCCESS;
		}
		cfp = gtcfile(s1);
		fprintf(cfp, "X %s %s!%s %s %s\n", file1, s2, file2, User, Optns);
		break;
	}
	return SUCCESS;
}

/*
 *	execute uux for remote uucp
 *
 *	return code - none
 */

xuux(ename, s1, f1, s2, f2, opts)
char *ename, *s1, *s2, *f1, *f2, *opts;
{
	char cmd[200];

	DEBUG(4, "Ropt(%s) ", Ropt);
	DEBUG(4, "ename(%s) ", ename);
	DEBUG(4, "s1(%s) ", s1);
	DEBUG(4, "f1(%s) ", f1);
	DEBUG(4, "s2(%s) ", s2);
	DEBUG(4, "f2(%s)\n", f2);
	sprintf(cmd, "uux %s %s!uucp %s %s!%s \\(%s!%s\\)",
	 Ropt, ename, opts,  s1, f1, s2, f2);
	DEBUG(4, "cmd (%s)\n", cmd);
	system(cmd);
	return;
}

FILE *Cfp = NULL;
char Cfile[NAMESIZE];

/*
 *	get a Cfile descriptor
 *
 *	return an open file descriptor
 */

FILE *
gtcfile(sys)
register char *sys;
{
	static char presys[8] = "";
	static int cmdcount = 0;
	register int savemask;

	if (strcmp(presys, sys) != SAME  /* this is !SAME on first call */
	  || Nbytes > MAXBYTES
	  || ++cmdcount > MAXCOUNT) {
		cmdcount = 1;
		Nbytes = 0;
		if (presys[0] != '\0') {
			clscfile();
		}
		gename(CMDPRE, sys, Grade, Cfile);
#ifdef VMS
		savemask = umask(~0600); /* vms must have read permission */
#else !VMS
		savemask = umask(~0200);
#endif !VMS
		Cfp = fopen(subfile(Cfile), "w");
		umask(savemask);
		if (Cfp == NULL) {
			syslog(LOG_WARNING, "fopen(%s) failed: %m",
				subfile(Cfile));
			cleanup(1);
		}
		strcpy(presys, sys);
	}
	return Cfp;
}

/*
 *	close cfile
 *
 *	return code - none
 */

clscfile()
{
	if (Cfp == NULL)
		return;
	fclose(Cfp);
	chmod(subfile(Cfile), ~WFMASK & 0777);
	logent(Cfile, "QUE'D");
	US_CRS(Cfile);
	Cfp = NULL;
}

/*
 * compile a list of all systems we are referencing
 */
chsys(s1)
register char *s1;
{
	if (xsflag < 0)
		xsflag = 0;
	else if (xsflag > 0)
		return;

	if (xsys[0] == '\0') {
		strncpy(xsys, s1, MAXBASENAME);
		return;
	}

	if (strncmp(xsys, s1, MAXBASENAME) == SAME)
		return;

	xsflag++;
	xsys[0] = '\0';
}
